### ISSUE TITLE
md-translate times out while waiting for translation

### ISSUE THREAD

**Issue by Occhioverde** [ilyachch/md_docs-trans-app#48]
Good morining.
As per the issue title, it seems that md-translate doesn't play nice with some Google Translate and Deepl site upgrades that have been rolled out since the last release.
After starting the program with  `md-translate [filename] -F en -T it -P google -N -v`, in fact, the Chrome window is opened and the first block of text is pasted into the webapp, but then is not retrieved when ready, with the subsequent crash with the following exception:
```
ERROR:md_translate.application:Error while translating file: [filename]
ERROR:md_translate.application:Message: 
Traceback (most recent call last):
  File "/home/rsacchetto/.local/pipx/venvs/md-translate/lib/python3.11/site-packages/md_translate/application.py", line 119, in process_file
    document.translate(provider)
  File "/home/rsacchetto/.local/pipx/venvs/md-translate/lib/python3.11/site-packages/md_translate/document/document.py", line 74, in translate
    translated_data = translator.translate(text=str(block))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rsacchetto/.local/pipx/venvs/md-translate/lib/python3.11/site-packages/md_translate/translators/_selenium_base.py", line 70, in translate
    self.wait_for_translation()
  File "/home/rsacchetto/.local/pipx/venvs/md-translate/lib/python3.11/site-packages/md_translate/translators/_selenium_base.py", line 149, in wait_for_translation
    self.WEBDRIVER_WAIT(self._driver, self.TRANSLATION_TIMEOUT).until(wait_for)
  File "/home/rsacchetto/.local/pipx/venvs/md-translate/lib/python3.11/site-packages/selenium/webdriver/support/wait.py", line 95, in until
    raise TimeoutException(message, screen, stacktrace)
selenium.common.exceptions.TimeoutException: Message:
```
The same situation happens with Deepl as the provider, with the only difference of a longer timeout time.
Unfortunately I'm absolutely ignorant in the field of web scraping and Python + Selenium programming, so I wasn't able to create a patch myself, but from a rapid skim through the source I was able to find that the CSS selectors md-translator was looking for to check whether the translations are ready are not valid anymore.
Maybe it is possible to replace them with a check that is less subject to changes, like something that waits for a web request to be completed?
In any case, thank you very much for your time and your work.

**Comment by ilyachch** [comment]
Thank you for your submission. I can confirm it.
The issue with deepl can be solved very easy - they just removed an attr, I was using for parsing. But they also made a constraint for 1500 symbols for free tier (Now I'm wondering how to get around this, or at least make some internal restrictions to notify the user about it)
The situation with google is harder - their markup is very hard to parse=(

I'll try to fix it - after it will be done, the new version will appear in PyPI, so, you'll need just to update the package

Also, If you know any good enough translating services or some other thoughts, feel free to write

**Comment by Occhioverde** [comment]
> I'll try to fix it - after it will be done, the new version will appear in PyPI

Thank you very much!

> But they also made a constraint for 1500 symbols for free tier

Maybe you can split the paragraphs to fit them in the 1500 symbols limit? Or, even better, you could add an option to login with a Premium Deepl account.

> The situation with google is harder - their markup is very hard to parse=(

Personally, that's not a big problem. I put Google Translate first in the Issue description because it's the most popular, but I actually prefer Deepl. Obviously, though, other people may not agree, so I can't pretend to speak for everyone.

> If you know any good enough translating services or some other thoughts, feel free to write

One service that may be worth noting is LibreTranslate, an open source translation API; it is still rough around the edges, but works pretty good and (unlike other providers) is a lot easier to integrate in third party apps. Maybe you can check it out.

**Comment by ilyachch** [comment]
> Maybe you can split the paragraphs to fit them in the 1500 symbols limit? Or, even better, you could add an option to login with a Premium Deepl account.

Me, as a Unix philosophy fan and python developer, have two rules - "The software should do one thing and do it good" from Unix and "Explicit is better than implicit" from Python. And I believe the software should not do any implicit changes to text. So, I think, the better way will be to do constraint with rising error, if deepl is used and there comes text bigger then 1000 symbols (this is an approximate value, because deepl cuts output, not input). The best way will be to make a separate provider for deepl API, with API key required in CLI args or settings file. And now I'm researching the way to implement it=)

> Personally, that's not a big problem. I put Google Translate first in the Issue description because it's the most popular, but I actually prefer Deepl. Obviously, though, other people may not agree, so I can't pretend to speak for everyone.

It my sound weird, but in my experience, the top of providers:

1) Deepl
2) Bing (surprise!)
3) Google
4) Yandex (as yandex is using google, but but it's more restrictive with antibot)

> One service that may be worth noting is [LibreTranslate](https://libretranslate.com/), an open source translation API; it is still rough around the edges, but works pretty good and (unlike other providers) is a lot easier to integrate in third party apps. Maybe you can check it out.

Yeah, I've heard about it and I have a plan to integrate it=)

**Comment by kingdase** [comment]
I fixed the function `click_cookies_accept` in deepl.py

Does someone has a patch for `check_for_translation` for deepl.py ?

```
~/github/md_docs-trans-app/md_translate/translators$ git diff deepl.py
diff --git a/md_translate/translators/deepl.py b/md_translate/translators/deepl.py
index 95e72e0..d4bc743 100644
--- a/md_translate/translators/deepl.py
+++ b/md_translate/translators/deepl.py
@@ -37,7 +37,7 @@ class DeeplTranslateProvider(SeleniumBaseTranslator):
     @safe_run(NoSuchElementException, default_return_value=None)
     def click_cookies_accept(self, btn_text: str) -> None:
         cookies_accept_button = self._driver.find_element(
-            by=self.WEBDRIVER_BY.CLASS_NAME, value='dl_cookieBanner--buttonAll'
+            by=self.WEBDRIVER_BY.CSS_SELECTOR, value='[data-testid="cookie-banner-strict-accept-all"]',
         )
         if cookies_accept_button:
             cookies_accept_button.click()

```

**Comment by ilyachch** [comment]
@Occhioverde Hi! Sorry for delay. The fixed version is released, you can update with `pipx update md-translate`. 

The Deepl and Google are working now (without constraints or new providers - I decided to do them as separate issues):

* https://github.com/ilyachch/md_docs-trans-app/issues/51
* https://github.com/ilyachch/md_docs-trans-app/issues/52
* https://github.com/ilyachch/md_docs-trans-app/issues/53
* https://github.com/ilyachch/md_docs-trans-app/issues/54

**Comment by Occhioverde** [comment]
> Hi! Sorry for delay.

Good morning! No problem, the last thing that I want is to put pressure!

> The Deepl and Google are working now

I had a spare minute, so I decided to go ahead and test the new version right away, but I noticed that, while the Google Translate provider works as expected, the Deepl one picks up the output a little too early, including the ellipses rendered while the translation process is running:

```
Many languages are dogmatic about the solutions they offer for handling
concurrent problems. For example, Erlang has elegant functionality for
message-passing concurrency but has only obscure ways to share state between
threads. Supporting only a subset of possible solutions is a reasonable
strategy for higher-level languages, because a higher-level language promises
benefits from giving up some control to gain abstractions. However, lower-level
languages are expected to provide the solution with the best performance in any
given situation and have fewer abstractions over the hardware. Therefore, Rust
offers a variety of tools for modeling problems in whatever way is appropriate
for your situation and requirements.

Molte lingue sono dogmatiche per quanto riguarda le soluzioni offerte per la gestione di
[...]
```

Since you said that you are currently working on the Deepl API integration, I don't know whether fixing this problem is worth your time (as it won't certainly be present in the API-powered provider), but if it's possible to write a quick patch it would be awesome!

Thanks again for your time and amazing work!

**Comment by ilyachch** [comment]
I can not predict the time to integrate API, but I think, it should not take much time.

As a temporary fix your case, I can suggest you to separate this paragraph into several smaller pieces (as translation is performed per paragraph) and, after translation, concatenate them again. I understand, that it is not looks like automated translation, but anyway, I believe it will take less time, then manual translation.

**Comment by ilyachch** [comment]
@Occhioverde 
Hey! I've just released new version with implementation of LibreTranslate and Deepl API integrations. 
So, you can try it - also, if you find any bugs, please, open an issue =)

**Comment by Occhioverde** [comment]
Good morning, @ilyachch.

Thank you very much for the updates! I tested the new API-powered implementation, and it seems to work like a charm! I'll try it more extensively along with a person who's working with me on the translation project for which md-translate was needed, and certainly let you know of any issue that should arise.

> The strange thing - I've just tested on text snippet, provided by you - and the whole text translated in single event.

Yeah, the issue seems to happen after some time, probably due to the overhead on the browser due to the multiple page reloads. Since that paragraph comes from the [Chapter 16 of the open-source Rust Book](https://github.com/rust-lang/book/blob/main/src/ch16-00-concurrency.md), maybe you can try translating the whole file to see whether you can reproduce the problem.

Also keep in mind that I'm running md-translate in a VM in order to keep Google Chrome isolated from my system and the fact that - in comparison to a bare-metal OS - VM resources are a little scarcer may have something to do with the longer loading times.
